### PR TITLE
History

### LINKED ISSUE #36: Historical data (server)
There should be a server which will store all the historical data with timestamps).

This is useful for debugging and analyzing the bots (e.q after night)

Those should be possible to visualize and if we have a play button to display events in high fps (like a movie) it would be perfect.

### PR DESCRIPTION
**Pull request by tivvit**
* stored in the main DB
* endpoint of the game server 
* able to return data for any season and tick
  * for the current season, it does not return commands (it does for the currently logged player - IMHO, it may be helpful also for the bots - may be used as external memory for previous decisions)

reports are missing - because they are not implemented yet.

closes #36

### CODE REVIEW

File: server/stycoon/history.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	reportUserId := int64(-1)
+	data, err := game.GetData(&reportUserId)
+	if err != nil {
+		return err
+	}
+	serializedData, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	_, err = game.db.Exec("insert into d_history (`season`, `tick`, `data`) values (?, ?, ?)", game.Tick.Season, game.Tick.Tick, serializedData)
```
Comment (malytomas) [inline server/stycoon/history.go:24]: season and tick is a primary key. multiple inserts will fail. it must be inserted all at once

File: database/database.sql [hunk 2]
```diff
@@ -2038,6 +2038,18 @@ CREATE TABLE IF NOT EXISTS `d_user_score` (
   CONSTRAINT `FK_d_user_score_d_user` FOREIGN KEY (`user`) REFERENCES `d_user` (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3;
 
+DROP TABLE IF EXISTS `d_history`;
+CREATE TABLE `d_history`
+(
+    `season`      int(11) NOT NULL,
+    `tick`        int(11) NOT NULL,
+    `reports`     longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL CHECK (json_valid(`reports`)),
```
Comment (malytomas) [inline database/database.sql:2046]: why nullable?

File: server/stycoon/history.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func FilterCommands(playerId *int64, ships *map[string]ShipsValue) {
+	for k, ship := range *ships {
+		if ship.Command == nil {
+			continue
+		}
+		shouldRemove := true
+		if playerId != nil && *playerId == ship.Player {
+			shouldRemove = false
```
Comment (malytomas) [inline server/stycoon/history.go:35]: invert the condition and use continue instead of variable

File: server/stycoon/history.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		filteredShip := ship
+		filteredShip.Command = nil
+		(*ships)[k] = filteredShip
+	}
+}
+
+func (game *Game) History(id HistoryIdentifier, playerId *int64) (HistoryEntry, error) {
+	var entry HistoryEntry
+	var staticData sql.NullString
+	err := game.db.QueryRow("select `static-data` from d_history where `season` = ? and `tick` = -1", id.Season).Scan(&staticData)
```
Comment (malytomas) [inline server/stycoon/history.go:49]: oh god i see it now. please split it into two tables instead.

Comment (malytomas) [reply server/stycoon/history.go:49]: one table for static data and second table for data and reports

File: server/stycoon/game.go [hunk 5]
```diff
@@ -165,7 +170,14 @@ func (game *Game) GetShips(playerId *int64, resources *map[int]map[string]*Tradi
 
 func (game *Game) getPlayerCommands(playerId int64) (map[int]Command, error) {
 	commands := make(map[int]Command)
-	rows, err := game.db.Query("select object.`id`, `type`, `target`, `resource`, `amount`, `class` from t_command join t_object object on t_command.ship = object.id where object.`owner` = ?", playerId)
+	var rows *sql.Rows
+	var err error
+	// return all commands for history report
+	if playerId == -1 {
+		rows, err = game.db.Query("select object.`id`, `type`, `target`, `resource`, `amount`, `class` from t_command join t_object object on t_command.ship = object.id")
```
Comment (malytomas) [inline server/stycoon/game.go:177]: put common part of the query in a constant/variable

**Comment by malytomas** [comment]
reports contain history too. they either have to be separated into individual ticks alongside the data. or have separate table with one row per season (updated after every tick in the current season)

**Comment by malytomas** [comment]
i can prepare the database for you if you want..

File: server/stycoon/history.go (hunk 4 above)
Comment (tivvit) [reply server/stycoon/history.go:49]: Ok, first I thought that this is simpler, but it is exactly the same (apart from lower readability). Changed

File: database/database.sql (hunk 2 above)
Comment (tivvit) [reply database/database.sql:2046]: explained in the following comments, removed (apart from reports which need to be solved later)

File: server/stycoon/history.go (hunk 1 above)
Comment (tivvit) [reply server/stycoon/history.go:24]: I count with that. It also depends on how we are going to store reports.

File: server/stycoon/history.go (hunk 3 above)
Comment (tivvit) [reply server/stycoon/history.go:35]: much nicer, thanks!

File: server/stycoon/game.go (hunk 5 above)
Comment (tivvit) [reply server/stycoon/game.go]: done - I agree with it OTOH syntax highlighting does not work for the string concat now :/

**Comment by tivvit** [comment]
> Member

I like the second solution. I will keep it in the same table with static data.

**Comment by tivvit** [comment]
> i can prepare the database for you if you want..

Thanks I can do it, it is simple in the end

**Comment by malytomas** [comment]
What changes would the visualization need to show historical data instead of current data? Is it as simple as changing url?

**Comment by tivvit** [comment]
agreed that we will add optional get params (tick, season) to the currently present endpoints instead of dedicated history endpoint

**Comment by malytomas** [comment]
please rename the tables to: d_history_data, d_history_reports, and d_history_static

File: database/database.sql [hunk 6]
```diff
@@ -3080,7 +3111,7 @@ CREATE TABLE IF NOT EXISTS `t_player` (
   `user` int(11) NOT NULL,
   `name` tinytext NOT NULL,
   `money` bigint(20) NOT NULL DEFAULT 5000000,
-  `color` tinytext NOT NULL DEFAULT '',
+  `color` tinytext NOT NULL DEFAULT '[0,0,0]' CHECK (json_valid(`color`)),
```
[the remark this answers is not in the document]
Comment (tivvit) [reply database/database.sql]: done

**Pull request merged**